# **Equipo 1**

Yocelin Juárez Arroyo - A00819828

Luis Arturo Dan Fong - A01650672

Octavio Edelberto Reyes Torres - A00269987


#Tener todo en un csv para trabajar desde aqui

In [44]:
import warnings
warnings.filterwarnings("ignore")

from google.colab import drive
drive.mount('/content/drive')
#folder_path = '/content/drive/My Drive/Proyecto Integrador/'
#folder_path = '/content/drive/My Drive/Proyecto Integrador/colab/'

#Octavio - Luis
#folder_path = '/content/drive/My Drive/Proyecto Integrador/'

#Yocelin (owner)
folder_path = '/content/drive/My Drive/Proyecto Integrador/Proyecto Integrador/'

# Importación de bibliotecas estándar
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import time
import gc

# Importación para visualización en Jupyter Notebook
from IPython.display import display

# Importación de preprocesamiento y modelos de scikit-learn
from sklearn.preprocessing import StandardScaler, PowerTransformer, LabelEncoder
from sklearn.decomposition import PCA
from sklearn.cluster import KMeans

# Importación de modelos de machine learning
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import RidgeClassifier, LogisticRegression
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis

# Importación para evaluación y ajuste de modelos
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import accuracy_score, f1_score, recall_score, classification_report, confusion_matrix


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [20]:
#Unir DF con una iniciativa Historia de éxito
ruta_archivo =  folder_path + 'bases/df_unido_limpioLGD_versionModelo.csv'
df_unidoLDG_encoded = pd.read_csv(ruta_archivo)

In [21]:
# Ajustar la configuración de pandas para mostrar todas las columnas
pd.set_option('display.max_columns', None)
# Mostrar las primeras filas del dataframe
display(df_unidoLDG_encoded.head())

categoria_dominio_correo_Educativo  categoria_dominio_correo_Genérico  \
0                               False                               True   
1                               False                               True   
2                               False                               True   
3                               False                               True   
4                               False                               True   

   categoria_dominio_correo_Otros  base_SEXO_Femenino  base_SEXO_Masculino  \
0                           False               False                 True   
1                           False               False                 True   
2                           False               False                 True   
3                           False               False                 True   
4                           False               False                 True   

   base_Tipo Escuela_Arquitectura y Diseño  \
0                                    False   
1                                    False   
2                                    False   
3                                    False   
4                                    False   

   base_Tipo Escuela_Gobierno y Ciencias Sociales  \
0                                           False   
1                                           False   
2                                           False   
3                                           False   
4                                           False   

   base_Tipo Escuela_Humanidades y Educación  \
0                                      False   
1                                      False   
2                                      False   
3                                      False   
4                                      False   

   base_Tipo Escuela_Ingeniería y Ciencias  \
0                                    False   
1                                    False   
2                                    False   
3                                    False   
4                                    False   

   base_Tipo Escuela_Medicina y Ciencias de la salud  \
0                                              False   
1                                              False   
2                                              False   
3                                              False   
4                                              False   

   base_Tipo Escuela_Negocios  base_IDCAMPUS_   base_IDCAMPUS_1  \
0                        True            False            False   
1                        True            False            False   
2                        True            False            False   
3                        True            False            False   
4                        True            False            False   

   base_IDCAMPUS_2  base_IDCAMPUS_8  base_IDCAMPUS_A  base_IDCAMPUS_A1  \
0            False            False             True             False   
1            False            False             True             False   
2            False            False            False             False   
3            False            False            False             False   
4            False            False            False             False   

   base_IDCAMPUS_AA  base_IDCAMPUS_C  base_IDCAMPUS_D  base_IDCAMPUS_E  \
0             False            False            False            False   
1             False            False            False            False   
2             False            False            False            False   
3             False            False            False            False   
4             False            False            False            False   

   base_IDCAMPUS_F  base_IDCAMPUS_G  base_IDCAMPUS_G1  base_IDCAMPUS_GA  \
0            False            False             False             False   
1            False            False             False             False   
2            False            False             False       

#**Avance 4. Modelos alternativos**

##6 Modelos

Se construyeron 6 modelos diferentes que ejecutamos para cada una de las variables objetivo definidas en nuestro proyecto (`'es_LDGincubadora'`, `'es_Voluntariado'`, `'es_HistoriaExito'`).

Evaluamos su rendimiento mediante métricas clave como **Accuracy**, **F1-score** y **Recall**, siendo Accuracy la métrica principal y considerando los tiempos de entrenamiento.

Los modelos utilizados son:

- **Naive Bayes**: Algoritmo de clasificación basado en el teorema de Bayes, ideal para problemas donde las características son independientes o casi independientes.

- **K-Nearest Neighbors (KNN)**: Clasificador que predice la clase de una observación en función de sus vecinos más cercanos en el espacio de características, útil en problemas donde la cercanía de los datos es relevante.

- **RidgeClassifier**: Clasificador lineal con regularización de Ridge (L2) que ayuda a controlar el sobreajuste al penalizar coeficientes altos, ideal para datos con colinealidad.

- **Linear Discriminant Analysis (LDA)**: Modelo estadístico que maximiza la separabilidad entre clases proyectando los datos en un espacio de menor dimensión, útil en problemas de clasificación.

- **Decision Tree**: Algoritmo de clasificación que utiliza una estructura de árbol para dividir los datos en función de características. Es fácil de interpretar y capaz de captar relaciones complejas.

- **Regresión Logística**: Modelo estadístico de clasificación que estima la probabilidad de que una observación pertenezca a una clase específica, ideal para problemas de clasificación binaria.

###Naive Bayes 3 variables objetivo

In [22]:
# Lista de variables objetivo
objective_variables = ['es_LDGincubadora', 'es_Voluntariado', 'es_HistoriaExito']

# DataFrame para almacenar resultados
results = []

# Iterar sobre cada variable objetivo
for objective in objective_variables:
    display(f"\nResultados para el modelo Naive Bayes con variable objetivo '{objective}':")

    # Separar las variables predictoras (X) y la variable objetivo (y)
    X = df_unidoLDG_encoded.drop(columns=objective_variables)
    y = df_unidoLDG_encoded[objective]

    # División de los datos en conjunto de entrenamiento y prueba
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

    # Medir el tiempo de entrenamiento
    start_time = time.time()

    # Crear y entrenar el modelo Naive Bayes
    model = GaussianNB()
    model.fit(X_train, y_train)

    # Calcular el tiempo de entrenamiento
    training_time = time.time() - start_time

    # Predicciones en el conjunto de prueba
    y_pred = model.predict(X_test)

    # Evaluar el rendimiento
    accuracy = accuracy_score(y_test, y_pred)
    f1 = f1_score(y_test, y_pred, average='weighted')
    recall = recall_score(y_test, y_pred, average='weighted')
    conf_matrix = confusion_matrix(y_test, y_pred)

    # Mostrar la matriz de confusión
    display("Matriz de Confusión:")
    display(pd.DataFrame(conf_matrix, index=['Clase 0', 'Clase 1'], columns=['Predicción 0', 'Predicción 1']))

    # Guardar los resultados en una lista
    results.append({
        "Variable Objetivo": objective,
        "Modelo": "Naive Bayes",
        "Accuracy": accuracy,
        "F1-Score": f1,
        "Recall": recall,
        "Tiempo de Entrenamiento (s)": training_time
    })

# Convertir los resultados a un DataFrame y mostrarlo
results_df = pd.DataFrame(results)
display("\nTabla Comparativa de Resultados Naive Bayes:")
display(results_df)

"\nResultados para el modelo Naive Bayes con variable objetivo 'es_LDGincubadora':"

'Matriz de Confusión:'

Predicción 0  Predicción 1
Clase 0         34627          4096
Clase 1          1312         27440

"\nResultados para el modelo Naive Bayes con variable objetivo 'es_Voluntariado':"

'Matriz de Confusión:'

Predicción 0  Predicción 1
Clase 0         29326          9283
Clase 1          1328         27538

"\nResultados para el modelo Naive Bayes con variable objetivo 'es_HistoriaExito':"

'Matriz de Confusión:'

Predicción 0  Predicción 1
Clase 0         35339          3042
Clase 1           227         28867

'\nTabla Comparativa de Resultados Naive Bayes:'

Variable Objetivo       Modelo  Accuracy  F1-Score    Recall  \
0  es_LDGincubadora  Naive Bayes  0.919852  0.920208  0.919852   
1   es_Voluntariado  Naive Bayes  0.842742  0.843233  0.842742   
2  es_HistoriaExito  Naive Bayes  0.951552  0.951748  0.951552   

   Tiempo de Entrenamiento (s)  
0                     2.581178  
1                     2.085361  
2                     2.443587

###KNN 3 variables objetivo

In [23]:
# Lista de variables objetivo
objective_variables = ['es_LDGincubadora', 'es_Voluntariado', 'es_HistoriaExito']

# DataFrame para almacenar resultados
results_knn = []

# Iterar sobre cada variable objetivo
for objective in objective_variables:
    display(f"\nResultados para el modelo KNN con variable objetivo '{objective}':")

    # Separar las variables predictoras (X) y la variable objetivo (y)
    X = df_unidoLDG_encoded.drop(columns=objective_variables)
    y = df_unidoLDG_encoded[objective]

    # División de los datos en conjunto de entrenamiento y prueba
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

    # Medir el tiempo de entrenamiento
    start_time = time.time()

    # Crear y entrenar el modelo KNN
    knn_model = KNeighborsClassifier(n_neighbors=3, algorithm='ball_tree')
    knn_model.fit(X_train, y_train)

    # Calcular el tiempo de entrenamiento
    training_time = time.time() - start_time

    # Predicciones en el conjunto de prueba
    y_pred = knn_model.predict(X_test)

    # Evaluar el rendimiento
    accuracy = accuracy_score(y_test, y_pred)
    f1 = f1_score(y_test, y_pred, average='weighted')
    recall = recall_score(y_test, y_pred, average='weighted')
    conf_matrix = confusion_matrix(y_test, y_pred)

    # Mostrar la matriz de confusión
    display("Matriz de Confusión:")
    display(pd.DataFrame(conf_matrix, index=['Clase 0', 'Clase 1'], columns=['Predicción 0', 'Predicción 1']))

    # Guardar los resultados en una lista
    results_knn.append({
        "Variable Objetivo": objective,
        "Modelo": "KNN",
        "Accuracy": accuracy,
        "F1-Score": f1,
        "Recall": recall,
        "Tiempo de Entrenamiento (s)": training_time
    })

# Convertir los resultados a un DataFrame y mostrarlo
results_knn_df = pd.DataFrame(results_knn)
display("\nTabla Comparativa de Resultados KNN:")
display(results_knn_df)

"\nResultados para el modelo KNN con variable objetivo 'es_LDGincubadora':"

'Matriz de Confusión:'

Predicción 0  Predicción 1
Clase 0         38718             5
Clase 1             1         28751

"\nResultados para el modelo KNN con variable objetivo 'es_Voluntariado':"

'Matriz de Confusión:'

Predicción 0  Predicción 1
Clase 0         38605             4
Clase 1             0         28866

"\nResultados para el modelo KNN con variable objetivo 'es_HistoriaExito':"

'Matriz de Confusión:'

Predicción 0  Predicción 1
Clase 0         38381             0
Clase 1             0         29094

'\nTabla Comparativa de Resultados KNN:'

Variable Objetivo Modelo  Accuracy  F1-Score    Recall  \
0  es_LDGincubadora    KNN  0.999911  0.999911  0.999911   
1   es_Voluntariado    KNN  0.999941  0.999941  0.999941   
2  es_HistoriaExito    KNN  1.000000  1.000000  1.000000   

   Tiempo de Entrenamiento (s)  
0                    24.874850  
1                    24.214987  
2                    25.614621

###RidgeClassifier 3 variables objetivo

In [24]:
# Lista de variables objetivo
objective_variables = ['es_LDGincubadora', 'es_Voluntariado', 'es_HistoriaExito']

# DataFrame para almacenar resultados
results_ridge = []

# Iterar sobre cada variable objetivo
for objective in objective_variables:
    display(f"\nResultados para el modelo RidgeClassifier con variable objetivo '{objective}':")

    # Separar las variables predictoras (X) y la variable objetivo (y)
    X = df_unidoLDG_encoded.drop(columns=objective_variables)
    y = df_unidoLDG_encoded[objective]

    # División de los datos en conjunto de entrenamiento y prueba
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

    # Medir el tiempo de entrenamiento
    start_time = time.time()

    # Crear y entrenar el modelo RidgeClassifier
    model = RidgeClassifier()
    model.fit(X_train, y_train)

    # Calcular el tiempo de entrenamiento
    training_time = time.time() - start_time

    # Predicciones en el conjunto de prueba
    y_pred = model.predict(X_test)

    # Evaluar el rendimiento
    accuracy = accuracy_score(y_test, y_pred)
    f1 = f1_score(y_test, y_pred, average='weighted')
    recall = recall_score(y_test, y_pred, average='weighted')
    conf_matrix = confusion_matrix(y_test, y_pred)

    # Mostrar la matriz de confusión
    display("Matriz de Confusión:")
    display(pd.DataFrame(conf_matrix, index=['Clase 0', 'Clase 1'], columns=['Predicción 0', 'Predicción 1']))

    # Guardar los resultados en una lista
    results_ridge.append({
        "Variable Objetivo": objective,
        "Modelo": "RidgeClassifier",
        "Accuracy": accuracy,
        "F1-Score": f1,
        "Recall": recall,
        "Tiempo de Entrenamiento (s)": training_time
    })

# Convertir los resultados a un DataFrame y mostrarlo
results_ridge_df = pd.DataFrame(results_ridge)
display("\nTabla Comparativa de Resultados RidgeClassifier:")
display(results_ridge_df)

"\nResultados para el modelo RidgeClassifier con variable objetivo 'es_LDGincubadora':"

'Matriz de Confusión:'

Predicción 0  Predicción 1
Clase 0         37761           962
Clase 1           794         27958

"\nResultados para el modelo RidgeClassifier con variable objetivo 'es_Voluntariado':"

'Matriz de Confusión:'

Predicción 0  Predicción 1
Clase 0         37333          1276
Clase 1           414         28452

"\nResultados para el modelo RidgeClassifier con variable objetivo 'es_HistoriaExito':"

'Matriz de Confusión:'

Predicción 0  Predicción 1
Clase 0         37914           467
Clase 1           232         28862

'\nTabla Comparativa de Resultados RidgeClassifier:'

Variable Objetivo           Modelo  Accuracy  F1-Score    Recall  \
0  es_LDGincubadora  RidgeClassifier  0.973976  0.973985  0.973976   
1   es_Voluntariado  RidgeClassifier  0.974954  0.974997  0.974954   
2  es_HistoriaExito  RidgeClassifier  0.989641  0.989646  0.989641   

   Tiempo de Entrenamiento (s)  
0                     3.624945  
1                     2.683274  
2                     3.468367

###Linear Discriminant Analysis (LDA) 3 variables objetivo

In [25]:
# Lista de variables objetivo
objective_variables = ['es_LDGincubadora', 'es_Voluntariado', 'es_HistoriaExito']

# DataFrame para almacenar resultados
results_lda = []

# Iterar sobre cada variable objetivo
for objective in objective_variables:
    display(f"\nResultados para el modelo LDA con variable objetivo '{objective}':")

    # Separar las variables predictoras (X) y la variable objetivo (y)
    X = df_unidoLDG_encoded.drop(columns=objective_variables)
    y = df_unidoLDG_encoded[objective]

    # División de los datos en conjunto de entrenamiento y prueba
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

    # Medir el tiempo de entrenamiento
    start_time = time.time()

    # Crear y entrenar el modelo LDA
    model = LinearDiscriminantAnalysis()
    model.fit(X_train, y_train)

    # Calcular el tiempo de entrenamiento
    training_time = time.time() - start_time

    # Predicciones en el conjunto de prueba
    y_pred = model.predict(X_test)

    # Evaluar el rendimiento
    accuracy = accuracy_score(y_test, y_pred)
    f1 = f1_score(y_test, y_pred, average='weighted')
    recall = recall_score(y_test, y_pred, average='weighted')
    conf_matrix = confusion_matrix(y_test, y_pred)

    # Mostrar la matriz de confusión
    display("Matriz de Confusión:")
    display(pd.DataFrame(conf_matrix, index=['Clase 0', 'Clase 1'], columns=['Predicción 0', 'Predicción 1']))

    # Guardar los resultados en una lista
    results_lda.append({
        "Variable Objetivo": objective,
        "Modelo": "LDA",
        "Accuracy": accuracy,
        "F1-Score": f1,
        "Recall": recall,
        "Tiempo de Entrenamiento (s)": training_time
    })

# Convertir los resultados a un DataFrame y mostrarlo
results_lda_df = pd.DataFrame(results_lda)
display("\nTabla Comparativa de Resultados LDA:")
display(results_lda_df)

"\nResultados para el modelo LDA con variable objetivo 'es_LDGincubadora':"

'Matriz de Confusión:'

Predicción 0  Predicción 1
Clase 0         37763           960
Clase 1           792         27960

"\nResultados para el modelo LDA con variable objetivo 'es_Voluntariado':"

'Matriz de Confusión:'

Predicción 0  Predicción 1
Clase 0         37330          1279
Clase 1           414         28452

"\nResultados para el modelo LDA con variable objetivo 'es_HistoriaExito':"

'Matriz de Confusión:'

Predicción 0  Predicción 1
Clase 0         37916           465
Clase 1           232         28862

'\nTabla Comparativa de Resultados LDA:'

Variable Objetivo Modelo  Accuracy  F1-Score    Recall  \
0  es_LDGincubadora    LDA  0.974035  0.974044  0.974035   
1   es_Voluntariado    LDA  0.974909  0.974952  0.974909   
2  es_HistoriaExito    LDA  0.989670  0.989675  0.989670   

   Tiempo de Entrenamiento (s)  
0                    25.588923  
1                    25.666720  
2                    27.098572

###Decision Tree 3 variables objetivo

In [45]:
# Lista de variables objetivo
objective_variables = ['es_LDGincubadora', 'es_Voluntariado', 'es_HistoriaExito']

# DataFrame para almacenar resultados
results_decision_tree = []

# Iterar sobre cada variable objetivo
for objective in objective_variables:
    display(f"\nResultados para el modelo Decision Tree con variable objetivo '{objective}':")

    # Separar las variables predictoras (X) y la variable objetivo (y)
    X = df_unidoLDG_encoded.drop(columns=objective_variables)
    y = df_unidoLDG_encoded[objective]

    # División de los datos en conjunto de entrenamiento y prueba
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

    # Medir el tiempo de entrenamiento
    start_time = time.time()

    # Crear y entrenar el modelo Decision Tree
    model = DecisionTreeClassifier(max_depth=5, min_samples_split=10, min_samples_leaf=5)
    model.fit(X_train, y_train)

    # Calcular el tiempo de entrenamiento
    training_time = time.time() - start_time

    # Predicciones en el conjunto de prueba
    y_pred = model.predict(X_test)

    # Evaluar el rendimiento
    accuracy = accuracy_score(y_test, y_pred)
    f1 = f1_score(y_test, y_pred, average='weighted')
    recall = recall_score(y_test, y_pred, average='weighted')
    conf_matrix = confusion_matrix(y_test, y_pred)

    # Mostrar la matriz de confusión
    display("Matriz de Confusión:")
    display(pd.DataFrame(conf_matrix, index=['Clase 0', 'Clase 1'], columns=['Predicción 0', 'Predicción 1']))

    # Guardar los resultados en una lista
    results_decision_tree.append({
        "Variable Objetivo": objective,
        "Modelo": "Decision Tree",
        "Accuracy": accuracy,
        "F1-Score": f1,
        "Recall": recall,
        "Tiempo de Entrenamiento (s)": training_time
    })

# Convertir los resultados a un DataFrame y mostrarlo
results_decision_tree_df = pd.DataFrame(results_decision_tree)
display("\nTabla Comparativa de Resultados Decision Tree:")
display(results_decision_tree_df)

"\nResultados para el modelo Decision Tree con variable objetivo 'es_LDGincubadora':"

'Matriz de Confusión:'

Predicción 0  Predicción 1
Clase 0         37326          1397
Clase 1          1329         27423

"\nResultados para el modelo Decision Tree con variable objetivo 'es_Voluntariado':"

'Matriz de Confusión:'

Predicción 0  Predicción 1
Clase 0         36883          1726
Clase 1          1021         27845

"\nResultados para el modelo Decision Tree con variable objetivo 'es_HistoriaExito':"

'Matriz de Confusión:'

Predicción 0  Predicción 1
Clase 0         38180           201
Clase 1           649         28445

'\nTabla Comparativa de Resultados Decision Tree:'

Variable Objetivo         Modelo  Accuracy  F1-Score    Recall  \
0  es_LDGincubadora  Decision Tree  0.959600  0.959606  0.959600   
1   es_Voluntariado  Decision Tree  0.959289  0.959347  0.959289   
2  es_HistoriaExito  Decision Tree  0.987403  0.987390  0.987403   

   Tiempo de Entrenamiento (s)  
0                     4.028357  
1                     4.817425  
2                     3.226993

###Regresión Logística 3 variables objetivo

In [27]:
# Lista de variables objetivo
objective_variables = ['es_LDGincubadora', 'es_Voluntariado', 'es_HistoriaExito']

# Tomar una muestra reducida del DataFrame para evitar carga excesiva
df_sample = df_unidoLDG_encoded.sample(frac=0.7, random_state=42)  # Reduce la muestra al 70% de los datos

# DataFrame para almacenar resultados
results_logistic = []

# Iterar sobre cada variable objetivo
for objective in objective_variables:
    display(f"\nResultados para el modelo de Regresión Logística con variable objetivo '{objective}':")

    # Separar las variables predictoras (X) y la variable objetivo (y)
    X = df_sample.drop(columns=objective_variables)
    y = df_sample[objective]

    # División de los datos en conjunto de entrenamiento y prueba
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

    # Medir el tiempo de entrenamiento
    start_time = time.time()

    # Crear y entrenar el modelo de Regresión Logística
    model = LogisticRegression(max_iter=500)
    model.fit(X_train, y_train)

    # Calcular el tiempo de entrenamiento
    training_time = time.time() - start_time

    # Predicciones en el conjunto de prueba
    y_pred = model.predict(X_test)

    # Evaluar el rendimiento
    accuracy = accuracy_score(y_test, y_pred)
    f1 = f1_score(y_test, y_pred, average='weighted')
    recall = recall_score(y_test, y_pred, average='weighted')
    conf_matrix = confusion_matrix(y_test, y_pred)

    # Mostrar la matriz de confusión
    display("Matriz de Confusión:")
    display(pd.DataFrame(conf_matrix, index=['Clase 0', 'Clase 1'], columns=['Predicción 0', 'Predicción 1']))

    # Guardar los resultados en una lista
    results_logistic.append({
        "Variable Objetivo": objective,
        "Modelo": "Logistic Regression",
        "Accuracy": accuracy,
        "F1-Score": f1,
        "Recall": recall,
        "Tiempo de Entrenamiento (s)": training_time
    })

# Convertir los resultados a un DataFrame y mostrarlo
results_logistic_df = pd.DataFrame(results_logistic)
display("\nTabla Comparativa de Resultados Logistic Regression:")
display(results_logistic_df)

"\nResultados para el modelo de Regresión Logística con variable objetivo 'es_LDGincubadora':"

'Matriz de Confusión:'

Predicción 0  Predicción 1
Clase 0         25899          1106
Clase 1          1405         18822

"\nResultados para el modelo de Regresión Logística con variable objetivo 'es_Voluntariado':"

'Matriz de Confusión:'

Predicción 0  Predicción 1
Clase 0         25501          1410
Clase 1          1483         18838

"\nResultados para el modelo de Regresión Logística con variable objetivo 'es_HistoriaExito':"

'Matriz de Confusión:'

Predicción 0  Predicción 1
Clase 0         26216           560
Clase 1           266         20190

'\nTabla Comparativa de Resultados Logistic Regression:'

Variable Objetivo               Modelo  Accuracy  F1-Score    Recall  \
0  es_LDGincubadora  Logistic Regression  0.946837  0.946785  0.946837   
1   es_Voluntariado  Logistic Regression  0.938749  0.938736  0.938749   
2  es_HistoriaExito  Logistic Regression  0.982512  0.982526  0.982512   

   Tiempo de Entrenamiento (s)  
0                    71.707996  
1                    71.590867  
2                    68.958340

##Comparativa de los 6 modelos

Extraímos los valores de accuracy, f1-score, recall y el tiempo de entrenamiento para cada modelo evaluado. Esto nos permitió analizar el rendimiento de cada uno en relación con las métricas principales.

Ordenamos la tabla según la métrica principal (accuracy) para identificar fácilmente los modelos más efectivos.

In [46]:
# Combinar todos los DataFrames de resultados en un solo DataFrame consolidado
all_results_df = pd.concat([
    results_df,                 # Resultados de Naive Bayes
    results_knn_df,             # Resultados de KNN
    results_ridge_df,           # Resultados de RidgeClassifier
    results_lda_df,             # Resultados de LDA
    results_decision_tree_df,   # Resultados de Decision Tree
    results_logistic_df         # Resultados de Logistic Regression
], ignore_index=True)

# Filtrar, ordenar por accuracy y mostrar resultados para cada variable objetivo por separado
for variable in all_results_df['Variable Objetivo'].unique():
    variable_df = all_results_df[all_results_df['Variable Objetivo'] == variable]
    variable_df = variable_df.sort_values(by='Accuracy', ascending=False)  # Ordenar por Accuracy en orden descendente
    display(f"\nResultados para la variable objetivo '{variable}' ordenados por Accuracy:")
    display(variable_df)

"\nResultados para la variable objetivo 'es_LDGincubadora' ordenados por Accuracy:"

Variable Objetivo               Modelo  Accuracy  F1-Score    Recall  \
3   es_LDGincubadora                  KNN  0.999911  0.999911  0.999911   
9   es_LDGincubadora                  LDA  0.974035  0.974044  0.974035   
6   es_LDGincubadora      RidgeClassifier  0.973976  0.973985  0.973976   
12  es_LDGincubadora        Decision Tree  0.959600  0.959606  0.959600   
15  es_LDGincubadora  Logistic Regression  0.946837  0.946785  0.946837   
0   es_LDGincubadora          Naive Bayes  0.919852  0.920208  0.919852   

    Tiempo de Entrenamiento (s)  
3                     24.874850  
9                     25.588923  
6                      3.624945  
12                     4.028357  
15                    71.707996  
0                      2.581178

"\nResultados para la variable objetivo 'es_Voluntariado' ordenados por Accuracy:"

Variable Objetivo               Modelo  Accuracy  F1-Score    Recall  \
4    es_Voluntariado                  KNN  0.999941  0.999941  0.999941   
7    es_Voluntariado      RidgeClassifier  0.974954  0.974997  0.974954   
10   es_Voluntariado                  LDA  0.974909  0.974952  0.974909   
13   es_Voluntariado        Decision Tree  0.959289  0.959347  0.959289   
16   es_Voluntariado  Logistic Regression  0.938749  0.938736  0.938749   
1    es_Voluntariado          Naive Bayes  0.842742  0.843233  0.842742   

    Tiempo de Entrenamiento (s)  
4                     24.214987  
7                      2.683274  
10                    25.666720  
13                     4.817425  
16                    71.590867  
1                      2.085361

"\nResultados para la variable objetivo 'es_HistoriaExito' ordenados por Accuracy:"

Variable Objetivo               Modelo  Accuracy  F1-Score    Recall  \
5   es_HistoriaExito                  KNN  1.000000  1.000000  1.000000   
11  es_HistoriaExito                  LDA  0.989670  0.989675  0.989670   
8   es_HistoriaExito      RidgeClassifier  0.989641  0.989646  0.989641   
14  es_HistoriaExito        Decision Tree  0.987403  0.987390  0.987403   
17  es_HistoriaExito  Logistic Regression  0.982512  0.982526  0.982512   
2   es_HistoriaExito          Naive Bayes  0.951552  0.951748  0.951552   

    Tiempo de Entrenamiento (s)  
5                     25.614621  
11                    27.098572  
8                      3.468367  
14                     3.226993  
17                    68.958340  
2                      2.443587

####Análisis de Resultados por Variable Objetivo

**Variable Objetivo: es_LDGincubadora**

El modelo KNN obtuvo la mayor precisión, con un accuracy de 0.999911 y un F1-Score de 0.999911, aunque presenta un tiempo de entrenamiento alto (24.52 segundos). Por otro lado, los modelos RidgeClassifier y LDA también muestran un buen desempeño, con una precisión de aproximadamente 0.974 y tiempos de entrenamiento significativamente menores, de 3.4 y 28.3 segundos, respectivamente. Como el tiempo de entrenamiento es un factor relevante, consideramos que **RidgeClassifier** es una excelente alternativa.

**Variable Objetivo: es_Voluntariado**

Nuevamente, el modelo KNN logra la mayor precisión, con un accuracy de 0.999941 y un F1-Score de 0.999941, pero su tiempo de entrenamiento es considerablemente alto (23.19 segundos). Tanto RidgeClassifier como LDA ofrecen precisiones similares, pero con tiempos de entrenamiento más bajos, especialmente RidgeClassifier, que solo requiere 2.47 segundos. Por eso **RidgeClassifier** sería nuestra elección.

**Variable Objetivo: es_HistoriaExito**

En esta variable, KNN alcanza una precisión perfecta, con un accuracy de 1.000000 y un F1-Score de 1.000000, pero con un tiempo de entrenamiento de 23.94 segundos. RidgeClassifier y LDA también muestran buenos resultados, con una precisión y tiempos de entrenamiento de aproximadamente 2.73 y 24.22 segundos, respectivamente. Por lo tanto, elegimos **RidgeClassifier**.



##Ajuste fino

Hemos decidido realizar el ajuste fino para `RidgeClassifier` en cada una de las variables objetivo.

####Ajuste Fino para es_LDGincubadora

Realizamos el ajuste fino de nuestro RidgeClassifier para optimizar su rendimiento en la variable objetivo 'es_LDGincubadora'. Para ello, dividimos los datos en entrenamiento y prueba, y usamos GridSearchCV para explorar las combinaciones de los hiperparámetros alpha (0.1, 1.0, 10.0), tol (1e-4, 1e-3, 1e-2), y solver ('auto', 'svd', 'cholesky') en una validación cruzada de 5 pliegues.

In [35]:
# Separar las variables predictoras (X) y la variable objetivo (y)
X = df_unidoLDG_encoded.drop(columns=['es_LDGincubadora', 'es_Voluntariado', 'es_HistoriaExito'])
y = df_unidoLDG_encoded['es_LDGincubadora']

# División de los datos en conjunto de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Definir los hiperparámetros a ajustar
param_grid = {
    'alpha': [0.1, 1.0, 10.0],
    'tol': [1e-4, 1e-3, 1e-2],
    'solver': ['auto', 'svd', 'cholesky']
}

# Configurar GridSearchCV
grid_search = GridSearchCV(RidgeClassifier(), param_grid, scoring='accuracy', cv=5, n_jobs=-1)

# Medir el tiempo de entrenamiento
start_time = time.time()

# Entrenar el modelo y realizar la búsqueda de hiperparámetros
grid_search.fit(X_train, y_train)

# Calcular el tiempo de entrenamiento
training_time = time.time() - start_time

# Mejor modelo encontrado
best_model = grid_search.best_estimator_

# Predicciones en el conjunto de prueba
y_pred = best_model.predict(X_test)

# Evaluar el rendimiento
accuracy = accuracy_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred, average='weighted')
recall = recall_score(y_test, y_pred, average='weighted')
conf_matrix = confusion_matrix(y_test, y_pred)

# Guardar los resultados en un DataFrame
results_ldgincubadora_df = pd.DataFrame([{
    "Variable Objetivo": 'es_LDGincubadora',
    "Modelo": "RidgeClassifier (Ajuste Fino)",
    "Mejores Hiperparámetros": grid_search.best_params_,
    "Accuracy": accuracy,
    "F1-Score": f1,
    "Recall": recall,
    "Tiempo de Entrenamiento (s)": training_time
}])

# Mostrar los resultados y la matriz de confusión
display("Matriz de Confusión:")
display(pd.DataFrame(conf_matrix, index=['Clase 0', 'Clase 1'], columns=['Predicción 0', 'Predicción 1']))
display("Resultados del ajuste fino para la variable objetivo 'es_LDGincubadora':")
display(results_ldgincubadora_df)

'Matriz de Confusión:'

Predicción 0  Predicción 1
Clase 0         37761           962
Clase 1           803         27949

"Resultados del ajuste fino para la variable objetivo 'es_LDGincubadora':"

Variable Objetivo                         Modelo  \
0  es_LDGincubadora  RidgeClassifier (Ajuste Fino)   

                            Mejores Hiperparámetros  Accuracy  F1-Score  \
0  {'alpha': 10.0, 'solver': 'auto', 'tol': 0.0001}  0.973842  0.973851   

     Recall  Tiempo de Entrenamiento (s)  
0  0.973842                  1018.107586

###Ajuste Fino para es_Voluntariado

Realizamos el ajuste fino del RidgeClassifier para la variable objetivo 'es_Voluntariado'. Dividimos los datos en entrenamiento y prueba y usamos GridSearchCV para explorar exhaustivamente todas las combinaciones de los hiperparámetros alpha (0.1, 1.0, 10.0), tol (1e-4, 1e-3, 1e-2), y solver ('auto', 'svd', 'cholesky') en una validación cruzada de 5 pliegues.

In [36]:
# Separar las variables predictoras (X) y la variable objetivo (y)
X = df_unidoLDG_encoded.drop(columns=['es_LDGincubadora', 'es_Voluntariado', 'es_HistoriaExito'])
y = df_unidoLDG_encoded['es_Voluntariado']

# División de los datos en conjunto de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Definir los hiperparámetros a ajustar
param_grid = {
    'alpha': [0.1, 1.0, 10.0],
    'tol': [1e-4, 1e-3, 1e-2],
    'solver': ['auto', 'svd', 'cholesky']
}

# Configurar GridSearchCV
grid_search = GridSearchCV(RidgeClassifier(), param_grid, scoring='accuracy', cv=5, n_jobs=-1)

# Medir el tiempo de entrenamiento
start_time = time.time()

# Entrenar el modelo y realizar la búsqueda de hiperparámetros
grid_search.fit(X_train, y_train)

# Calcular el tiempo de entrenamiento
training_time = time.time() - start_time

# Mejor modelo encontrado
best_model = grid_search.best_estimator_

# Predicciones en el conjunto de prueba
y_pred = best_model.predict(X_test)

# Evaluar el rendimiento
accuracy = accuracy_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred, average='weighted')
recall = recall_score(y_test, y_pred, average='weighted')
conf_matrix = confusion_matrix(y_test, y_pred)

# Guardar los resultados en un DataFrame
results_voluntariado_df = pd.DataFrame([{
    "Variable Objetivo": 'es_Voluntariado',
    "Modelo": "RidgeClassifier (Ajuste Fino)",
    "Mejores Hiperparámetros": grid_search.best_params_,
    "Accuracy": accuracy,
    "F1-Score": f1,
    "Recall": recall,
    "Tiempo de Entrenamiento (s)": training_time
}])

# Mostrar los resultados y la matriz de confusión
display("Matriz de Confusión:")
display(pd.DataFrame(conf_matrix, index=['Clase 0', 'Clase 1'], columns=['Predicción 0', 'Predicción 1']))
display("Resultados del ajuste fino para la variable objetivo 'es_Voluntariado':")
display(results_voluntariado_df)

'Matriz de Confusión:'

Predicción 0  Predicción 1
Clase 0         37341          1268
Clase 1           411         28455

"Resultados del ajuste fino para la variable objetivo 'es_Voluntariado':"

Variable Objetivo                         Modelo  \
0   es_Voluntariado  RidgeClassifier (Ajuste Fino)   

                            Mejores Hiperparámetros  Accuracy  F1-Score  \
0  {'alpha': 10.0, 'solver': 'auto', 'tol': 0.0001}  0.975117  0.975159   

     Recall  Tiempo de Entrenamiento (s)  
0  0.975117                   953.213034

### Ajuste fino para es_HistoriaExito

Realizamos el ajuste fino del RidgeClassifier para la variable objetivo 'es_HistoriaExito'. Dividimos los datos en entrenamiento y prueba y utilizamos GridSearchCV para explorar todas las combinaciones de los hiperparámetros alpha (0.1, 1.0, 10.0), tol (1e-4, 1e-3, 1e-2), y solver ('auto', 'svd', 'cholesky') mediante validación cruzada de 5 pliegues.

In [37]:
# Separar las variables predictoras (X) y la variable objetivo (y)
X = df_unidoLDG_encoded.drop(columns=['es_LDGincubadora', 'es_Voluntariado', 'es_HistoriaExito'])
y = df_unidoLDG_encoded['es_HistoriaExito']

# División de los datos en conjunto de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Definir los hiperparámetros a ajustar
param_grid = {
    'alpha': [0.1, 1.0, 10.0],
    'tol': [1e-4, 1e-3, 1e-2],
    'solver': ['auto', 'svd', 'cholesky']
}

# Configurar GridSearchCV
grid_search = GridSearchCV(RidgeClassifier(), param_grid, scoring='accuracy', cv=5, n_jobs=-1)

# Medir el tiempo de entrenamiento
start_time = time.time()

# Entrenar el modelo y realizar la búsqueda de hiperparámetros
grid_search.fit(X_train, y_train)

# Calcular el tiempo de entrenamiento
training_time = time.time() - start_time

# Mejor modelo encontrado
best_model = grid_search.best_estimator_

# Predicciones en el conjunto de prueba
y_pred = best_model.predict(X_test)

# Evaluar el rendimiento
accuracy = accuracy_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred, average='weighted')
recall = recall_score(y_test, y_pred, average='weighted')
conf_matrix = confusion_matrix(y_test, y_pred)

# Guardar los resultados en un DataFrame
results_historiaexito_df = pd.DataFrame([{
    "Variable Objetivo": 'es_HistoriaExito',
    "Modelo": "RidgeClassifier (Ajuste Fino)",
    "Mejores Hiperparámetros": grid_search.best_params_,
    "Accuracy": accuracy,
    "F1-Score": f1,
    "Recall": recall,
    "Tiempo de Entrenamiento (s)": training_time
}])

# Mostrar los resultados y la matriz de confusión
display("Matriz de Confusión:")
display(pd.DataFrame(conf_matrix, index=['Clase 0', 'Clase 1'], columns=['Predicción 0', 'Predicción 1']))
display("Resultados del ajuste fino para la variable objetivo 'es_HistoriaExito':")
display(results_historiaexito_df)

'Matriz de Confusión:'

Predicción 0  Predicción 1
Clase 0         37916           465
Clase 1           232         28862

"Resultados del ajuste fino para la variable objetivo 'es_HistoriaExito':"

Variable Objetivo                         Modelo  \
0  es_HistoriaExito  RidgeClassifier (Ajuste Fino)   

                           Mejores Hiperparámetros  Accuracy  F1-Score  \
0  {'alpha': 0.1, 'solver': 'auto', 'tol': 0.0001}   0.98967  0.989675   

    Recall  Tiempo de Entrenamiento (s)  
0  0.98967                   951.634447

##Comparación de Resultados

In [43]:
# Obtenemos los valores de los hiperparámetros por defecto de RidgeClassifier de la documentación oficial de scikit-learn y del código fuente de la biblioteca
# Definir los hiperparámetros por defecto de RidgeClassifier
default_hyperparameters = {'alpha': 1.0, 'solver': 'auto', 'tol': 0.001}

# Concatenar los DataFrames de resultados antes y después del ajuste fino para RidgeClassifier
for variable in all_results_df['Variable Objetivo'].unique():
    # Filtrar los resultados antes del ajuste fino
    before_fine_tuning_df = all_results_df[(all_results_df['Variable Objetivo'] == variable) & (all_results_df['Modelo'] == "RidgeClassifier")]

    # Filtrar los resultados después del ajuste fino usando los DataFrames correspondientes
    if variable == 'es_LDGincubadora':
        after_fine_tuning_df = results_ldgincubadora_df
    elif variable == 'es_Voluntariado':
        after_fine_tuning_df = results_voluntariado_df
    elif variable == 'es_HistoriaExito':
        after_fine_tuning_df = results_historiaexito_df

    # Concatenar ambos DataFrames
    comparative_df = pd.concat([before_fine_tuning_df, after_fine_tuning_df], ignore_index=True)

    # Asignar los hiperparámetros por defecto donde haya valores NaN en "Mejores Hiperparámetros"
    comparative_df['Mejores Hiperparámetros'] = comparative_df['Mejores Hiperparámetros'].apply(
        lambda x: default_hyperparameters if pd.isna(x) else x
    )

    # Eliminar la columna de tiempo de entrenamiento
    comparative_df = comparative_df.drop(columns=["Tiempo de Entrenamiento (s)"], errors='ignore')

    # Mostrar los resultados comparativos para cada variable objetivo
    display(f"\nComparación de resultados para la variable objetivo '{variable}':")
    display(comparative_df)

"\nComparación de resultados para la variable objetivo 'es_LDGincubadora':"

Variable Objetivo                         Modelo  Accuracy  F1-Score  \
0  es_LDGincubadora                RidgeClassifier  0.973976  0.973985   
1  es_LDGincubadora  RidgeClassifier (Ajuste Fino)  0.973842  0.973851   

     Recall                           Mejores Hiperparámetros  
0  0.973976    {'alpha': 1.0, 'solver': 'auto', 'tol': 0.001}  
1  0.973842  {'alpha': 10.0, 'solver': 'auto', 'tol': 0.0001}

"\nComparación de resultados para la variable objetivo 'es_Voluntariado':"

Variable Objetivo                         Modelo  Accuracy  F1-Score  \
0   es_Voluntariado                RidgeClassifier  0.974954  0.974997   
1   es_Voluntariado  RidgeClassifier (Ajuste Fino)  0.975117  0.975159   

     Recall                           Mejores Hiperparámetros  
0  0.974954    {'alpha': 1.0, 'solver': 'auto', 'tol': 0.001}  
1  0.975117  {'alpha': 10.0, 'solver': 'auto', 'tol': 0.0001}

"\nComparación de resultados para la variable objetivo 'es_HistoriaExito':"

Variable Objetivo                         Modelo  Accuracy  F1-Score  \
0  es_HistoriaExito                RidgeClassifier  0.989641  0.989646   
1  es_HistoriaExito  RidgeClassifier (Ajuste Fino)  0.989670  0.989675   

     Recall                          Mejores Hiperparámetros  
0  0.989641   {'alpha': 1.0, 'solver': 'auto', 'tol': 0.001}  
1  0.989670  {'alpha': 0.1, 'solver': 'auto', 'tol': 0.0001}

**Variable Objetivo: es_LDGincubadora**

Antes del ajuste fino, el modelo obtuvo un Accuracy de 0.973976, un F1-Score de 0.973985, y un Recall de 0.973976 usando los hiperparámetros por defecto (alpha=1.0, solver='auto', tol=0.001). Después del ajuste fino, los valores disminuyeron ligeramente en todas las métricas, al aplicar la configuración alpha=10.0, solver='auto', tol=0.0001. Dado que el ajuste fino no mejoró el rendimiento, mantendremos los hiperparámetros por defecto para esta variable.

**Variable Objetivo: es_Voluntariado**

Para esta variable, el modelo con los valores por defecto alcanzó un Accuracy de aproximadamente 0.97495, con resultados muy similares para F1-Score y Recall. Tras el ajuste fino, estas métricas aumentaron ligeramente usando alpha=10.0, solver='auto', tol=0.0001. Aunque la mejora es pequeña, utilizaremos los hiperparámetros ajustados para maximizar el rendimiento de este modelo.

**Variable Objetivo: es_HistoriaExito**

Para es_HistoriaExito, el rendimiento del modelo fue excelente tanto antes como después del ajuste fino, con una Accuracy inicial de 0.989641 que aumentó a 0.989670 después del ajuste fino. La configuración ajustada (alpha=0.1, solver='auto', tol=0.0001) aportó una mejora leve, por lo que optaremos por mantener estos hiperparámetros ajustados para esta variable.

##Modelo individual final para cada variable objetivo

Concluimos que el **RidgeClassifier** es la elección óptima para nuestras variables objetivo (es_LDGincubadora, es_Voluntariado, y es_HistoriaExito) debido a su balance entre precisión y eficiencia en tiempos de entrenamiento.

A pesar de que el modelo KNN mostró la mayor precisión, su tiempo de entrenamiento significativamente alto hace que no sea la opción más viable en términos de costo computacional. Por su parte, RidgeClassifier demostró ser una excelente alternativa, con un desempeño muy competitivo en precisión y métricas de rendimiento, y tiempos de entrenamiento mucho más bajos.

Para cada variable objetivo, realizamos un ajuste fino de RidgeClassifier explorando combinaciones de los hiperparámetros **alpha, tol, y solver**. Estos ajustes nos permitieron optimizar el rendimiento del modelo de manera precisa para cada variable:

**Variable Objetivo: es_LDGincubadora**
* El ajuste fino no aportó mejoras significativas sobre los hiperparámetros por defecto **(alpha=1.0, solver='auto', tol=0.001)**, por lo que optamos por conservar esta configuración.

**Variable Objetivo: es_Voluntariado**
* En este caso, el ajuste fino logró una pequeña mejora en precisión y F1-Score con la configuración **alpha=10.0, solver='auto', tol=0.0001**.

**Variable Objetivo: es_HistoriaExito**
* Para es_HistoriaExito, el ajuste fino con **alpha=0.1, solver='auto', tol=0.0001** aumentó levemente el rendimiento del modelo.